src/dataflow/src/render/join.rs

```diff
@@ -46,46 +50,90 @@ where
+
+            // Determine dummy columns for un-demanded outputs, and a projection.
+            let (dummies, demand_projection) = if let Some(demand) = demand {
+                let mut dummies = Vec::new();
+                let mut demand_projection = Vec::new();
+                for (column, typ) in relation_expr.typ().column_types.into_iter().enumerate() {
+                    if demand.contains(&column) {
+                        demand_projection.push(column);
+                    } else {
+                        demand_projection.push(output_arity + dummies.len());
+                        dummies.push(ScalarExpr::literal_ok(Datum::Dummy, typ));
+                    }
+                }
+                (dummies, demand_projection)
+            } else {
+                (Vec::new(), (0..output_arity).collect::<Vec<_>>())
+            };
 
+            let (map, filter, project) = map_filter_project.as_map_filter_project();
+
+            let map_filter_project = MapFilterProject::new(output_arity)
+                .map(dummies)
+                .project(demand_projection)
+                .map(map)
+                .filter(filter)
+                .project(project);
+
+            // Other than the stream of updates, our loop-carried state are these
+            // three variables: `column_map`, `equivalences`, and `mfp`, which
+            // record the current locations of extended output columns, and what
+            // work remains to be done on them (other than the joining itself).
+            let mut column_map = HashMap::new();
+            for column in input_mapper.global_columns(*start) {
+                column_map.insert(column, column_map.len());
+            }
             // We maintain a private copy of `equivalences`, which we will digest
             // as we produce the join.
             let mut equivalences = equivalences.clone();
             for equivalence in equivalences.iter_mut() {
                 equivalence.sort();
                 equivalence.dedup();
             }
-
-            let input_mapper = expr::JoinInputMapper::new(inputs);
-
-            // Unwrap demand
-            // TODO: If we pushed predicates into the operator, we could have a
-            // more accurate view of demand that does not include the support of
-            // all predicates.
-            let demand = demand.clone().unwrap_or_else(|| (0..arity).collect());
+            // We maintain a private copy of `map_filter_project`, which we will
+            // digest as we produce the join.
+            let mut mfp = map_filter_project;
 
             // This collection will evolve as we join in more inputs.
+            // TODO(mcsherry): this clashes with the subsequent arrangement-based input.
+            // TODO(mcsherry): determine and apply closure here in `flat_map_ref` form.
+            // TODO(mcsherry): If we plan to use an arrangement, should one exist, then
```

Comment:
are the todos on lines 102 and 104 saying the same thing or different things? hard to tell from the text but it seems like they are saying the same thing twice (nbd just want to clarify)

Reply:
I think that is right. The first one I think decorated some different code with a less clear complaint (I think it used to be one a different test for whether we should use the arrangement).